### PR TITLE
Support Fee Items, custom Payment Accounts in QBO Invoice Sync

### PR DESCRIPTION
**Pull request by ananelson**
Note: I have a little more work to do on this in the next 24 hours. I want to test if fee_amount is > 0 before generating a line item, and I separately want to add support for QBO Classes.

Opening this now to start the feedback process.

### Fix Logging Issue

Unrelated to other changes, address issue with QBO API's logging. API logs to /tmp by default which may not be writable in every system. Allow deactivation of logs, and allow specifying an alternate log location.

### Fee Items

Our accounting practice is to create line items on invoices to reflect merchant/banking fees. This results in the total amount of the transaction being less than the nominal amount. For example, with a $5 donation with $0.41 going to merchant fees, the transactions should look like this:

image

Now if the Checkbox for Add Fee Lines? is checked, then a line item corresponding to fee amount is generated. Furthermore, this amount is also deducted from the payment total reflecting the actual amount which will be credited.

Separately, the Use Deposit Accounts? option will credit payments received to the account specified for the Payment Processor, rather than Undeposited Funds.

### CODE REVIEW

File: CRM/Civiquickbooks/Form/Settings.php [hunk 1]
```diff
@@ -139,10 +139,22 @@ public function getFormSettings() {
...
+        if ($setting_info['type'] == "Boolean" && !array_key_exists($setting_name, $values)) {
+            $values[$setting_name] = 0;
+        }
+    }
+
     $previousValues = CRM_Quickbooks_APIHelper::getQuickBooksCredentials();
 
     civicrm_api3('setting', 'create', $values);
 
+    // FIXME this erases valid access tokens any time settings are saved
```
Comment (ananelson) [inline CRM/Civiquickbooks/Form/Settings.php:157]: I want to fix this but want to understand what the scenario is that this if statement is allowing for.

File: CRM/Civiquickbooks/Invoice.php [hunk 2]
```diff
@@ -627,7 +708,7 @@ protected function mapToAccounts($db_contribution, $accountsID, $SyncToken, $qb_
       $new_invoice += array(
         'TxnDate' => $receive_date,
         'DueDate' => $due_date,
-        'DocNumber' => 'Civi-' . $db_contribution['id'],
+        'DocNumber' => $invoice_prefix . sprintf("%06d", $db_contribution['id']),
```
Comment (ananelson) [inline CRM/Civiquickbooks/Invoice.php:711]: Default invoice number will now look like: Civi-000001

I like my invoice numbers with leading zeros for better sortability. I can make this an option if you don't want it for everyone. Also now the prefix is easy to specify. (ours is CIV-)

**Comment by ananelson** [comment]
Also - per Mattermost - adding checkboxes has introduced a warning I can't figure out how to fix.

**Comment by agileware-justin** [comment]
@ananelson thanks for work and taking the time to submit a PR.

CiviCRM project does have a few guidelines which are worth reviewing as they are relevant here, please read:
https://docs.civicrm.org/dev/en/latest/tools/git/#pr-scope
https://docs.civicrm.org/dev/en/latest/tools/git/#rebasing-interactive (as much as I hate this one because always do it wrong)

I think what you really want to do is submit multiple PRs to propose discrete improvemnts or bug fixes. Rather than one big PR.

Rebasing here is also a good idea - show us your final work, not all the working drafts AND final work.

Did you want to have another crack at it?

**Pull request closed without merging**